plugins/intel-usb4/fu-intel-usb4-device.c

```diff
@@ -429,196 +421,131 @@ fu_dell_dock_usb4_activate(FuDevice *device, FuProgress *progress, GError **erro
 	g_return_val_if_fail(FU_IS_FIRMWARE(firmware), FALSE);
 
-	/* get default image */
-	fw_image = fu_firmware_get_bytes(firmware, error);
+	/* get payload */
+	fw_image = fu_firmware_get_image_by_id_bytes(firmware, FU_FIRMWARE_ID_PAYLOAD, error);
 	if (fw_image == NULL)
 		return FALSE;
 
-	fw_buf = g_bytes_get_data(fw_image, &fw_blob_size);
-	g_debug("total image size: %" G_GSIZE_FORMAT, fw_blob_size);
-
-	/* get header offset */
-	tmp_header = (guint8 *)&fw_header_offset;
-	if (!fu_memcpy_safe(tmp_header,
-			    sizeof(guint32),
-			    0x0,
-			    fw_buf,
-			    fw_blob_size,
-			    0x0,
-			    sizeof(guint32),
-			    error))
-		return FALSE;
-
-	g_debug("image header size: %" G_GUINT32_FORMAT, fw_header_offset);
-	if (fw_header_offset > fw_blob_size) {
-		g_set_error(error,
-			    FWUPD_ERROR,
-			    FWUPD_ERROR_INVALID_FILE,
-			    "image header is too big: %" G_GUINT32_FORMAT,
-			    fw_header_offset);
-		return FALSE;
-	}
-
-	/* get firmware version, vendor-id, product-id */
-	fw_version = g_strdup_printf("%02x.%02x",
-				     fw_buf[fw_header_offset + NVM_VER_OFFSET_MAJOR],
-				     fw_buf[fw_header_offset + NVM_VER_OFFSET_MINOR]);
-	fw_vendor_id = g_strdup_printf("%02x%02x",
-				       fw_buf[fw_header_offset + NVM_VID_OFFSET_MAJOR],
-				       fw_buf[fw_header_offset + NVM_VID_OFFSET_MINOR]);
-	fw_product_id = g_strdup_printf("%02x%02x",
-					fw_buf[fw_header_offset + NVM_PID_OFFSET_MAJOR],
-					fw_buf[fw_header_offset + NVM_PID_OFFSET_MINOR]);
-
-	g_debug("writing Thunderbolt firmware version %s", fw_version);
-	g_debug("writing Thunderbolt product-id %s", fw_product_id);
-	g_debug("writing Thunderbolt vendor-id %s", fw_vendor_id);
-
-	/* compare vendor-id, product-id between firmware blob and NVM */
-	if (!fu_dell_dock_usb4_hub_nvm_read(device, nvm_buf, NVM_READ_LENGTH, 0, error)) {
-		g_prefix_error(error, "NVM READ error: ");
-		return FALSE;
-	}
-	nvm_vendor_id = g_strdup_printf("%02x%02x",
-					nvm_buf[NVM_VID_OFFSET_MAJOR],
-					nvm_buf[NVM_VID_OFFSET_MINOR]);
-	nvm_product_id = g_strdup_printf("%02x%02x",
-					 nvm_buf[NVM_PID_OFFSET_MAJOR],
-					 nvm_buf[NVM_PID_OFFSET_MINOR]);
-
-	if (((flags & FWUPD_INSTALL_FLAG_FORCE) == 0) &&
-	    (g_strcmp0(nvm_vendor_id, fw_vendor_id) != 0 ||
-	     g_strcmp0(nvm_product_id, fw_product_id) != 0)) {
-		g_set_error(error,
-			    FWUPD_ERROR,
-			    FWUPD_ERROR_NOT_SUPPORTED,
-			    "Thunderbolt firmware vendor_id %s, product_id %s"
-			    "doesn't match NVM vendor_id %s, product_id %s",
-			    fw_vendor_id,
-			    fw_product_id,
-			    nvm_vendor_id,
-			    nvm_product_id);
-		return FALSE;
-	}
-
 	/* firmware install */
-	fw_buf += fw_header_offset;
-	fw_blob_size -= fw_header_offset;
-	if (!fu_dell_dock_usb4_hub_nvm_write(device, fw_buf, fw_blob_size, 0, progress, error))
+	if (!fu_intel_usb4_device_nvm_write(device, fw_image, 0, progress, error))
 		return FALSE;
 
+	/* success */
 	fu_device_add_flag(device, FWUPD_DEVICE_FLAG_NEEDS_ACTIVATION);
```

Comment:
In Dell dock this is handled by the dell-dock plugin. In other implementations it might not be true and you need to explicitly activate or another plugin needs to do it.

Reply:
> In other implementations it might not be true

Surely all Goshen ridge devices will need `OP_NVM_AUTH_WRITE` - i.e. NVM authenticate?

Reply:
Dell dock splits the steps and runs authentication when unplugged or when someone manually authenticates.  I would expect when this plugin is used alone with a Goshen Ridge that everything is in one swoop.

Reply:
Currently dell dock says "this is **my** goshen ridge" by looking at the VID and PID. Maybe we should add the parent VID/PID and match that instead?

Reply:
Well what I was thinking was in this plugin set it up as though it's not working with Dell dock. Then let Dell dock add or remove relevant flags.

That's how Dell dock interacted with thunderbolt plugin too.

So the idea is you can run this plugin alone to update the Goshen ridge or you can use Dell dock together with it and do the full Dell dock flow.

Reply:
So fu_intel_usb4_device_write_firmware() would call activate automatically if FWUPD_DEVICE_FLAG_NEEDS_ACTIVATION() wasn't set by dell-dock?

Reply:
Yes, exactly.

Reply:
if so, FWUPD_DEVICE_FLAG_NEEDS_ACTIVATION is probably the wrong thing, and we want to do this with something like FWUPD_DEVICE_FLAG_USABLE_DURING_UPDATE instead.

Reply:
Yeah that makes sense to me too.  I'm pretty sure that's what we use in `plugins/thunderbolt` too.

Reply:
Ahh so FWUPD_DEVICE_FLAG_SKIPS_RESTART is what we're using in tbt, and I can use that too -- but how do we know that the usb4 device that shows up in fu_plugin_dell_dock_device_registered() is actually a Dell GDR and not just a GR controller from another vendor? They're the same VID/PID for all OEMs.

Reply:
The GUID that's built for USB4 is from the vendor in the NVM isn't it?  Not from the USB VID/PID.

Reply:
Nope, it's the USB VID/PID. I guess we could read back the NVM VEN/DEV -- what do we expect that to be for the Dell dock? If there's any public firmware I can read it back and answer the question myself.

Reply:
Aha, so FuDellDockUsb4 checks the NVM by checking DELL_DOCK_USB4_INSTANCE_ID in ->setup() -- that makes sense now. Sorry for the noise. I've simplified what we're changing in dell-dock -- any other fixups or changes we can do in a different PR -- I want to keep these changes as minimal as possible. I've pushed this -- re-review most welcome.